A43-sdk-authorization.md

```diff
@@ -0,0 +1,713 @@
+  									  "headers (e.g. those listed in \"Connection\""
+  									  "header), HTTP/2 pseudo headers (\":\"-prefixed)"
+  									  "and headers prefixed with \"grpc-\".",
+  									"type": "string"
+  								},
+  								"values": {
+  									"description": "A list of header values to"
+  									  "match. The header values are ORed together,"
+  									  "i.e., it matches if one of the values"
+  									  "matches. Multi-valued headers are considered"
+  									  "a single value with commas added between"
+  									  "values."
+  									  "This field supports Exact, Prefix, Suffix"
+  									  "and Presence match."
+  									  "- Exact match: \"abc\" will match on value"
+  									  "  \"abc\"."
+  									  "- Prefix match: \"abc*\" will match on value"
+  									  "  \"abc\" and \"abcd\"."
+  									  "- Suffix match: \"*abc\" will match on value"
+  									  "  \"abc\" and \"xabc\"."
+  									  "- Presence match: \"*\" will match when the"
+  									  "  value is not empty.",
+  									"type": "array",
+  									"items": {
+  										"type": "string"
+  									}
+  								}
+  							},
+  							"required": ["key", "values"]
+  						}
+  					}
+  				}
+  			}
+  		},
+  		"required": ["name"]
+  	}
+  },
+  "description": "AuthorizationPolicy defines which principals are permitted to"
+    "access which resource. Resources are RPC methods scoped by services.",
+  "type": "object",
+  "properties": {
+  	"name": {
+  		"description": "The name of an authorization policy. It is mainly for"
+  		  "monitoring and error message generation.",
+  		"type": "string"
+  	},
+  	"deny_rules": {
+  		"description": "List of deny rules to match. If a request matches any of the"
+  		  "deny rules, then it will be denied. If none of the deny rules matches or"
+  		  "there are no deny rules, the allow rules will be evaluated.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	},
+  	"allow_rules": {
+  		"description": "List of allow rules to match. The allow rules will only be"
+  		  "evaluated after the deny rules. If a request matches any of the allow"
+  		  "rules, then it will allowed. If none of the allow rules matches, it will"
+  		  "be denied.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	}
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+#### Details
+
+gRPC SDK authorization policy has a list of *allow rules* and a list of *deny rules*.
+The following sequence is followed to make an authorization decision -
+1. Check for a match in *deny rules*, if a match is found, the request will be denied.
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in *allow rules*, if a match is found, the request will be allowed.
+   Note that allow rules is a required field here, if not present, policy is invalid.
```

Comment:
`allow_rules` is a required field but its value can be an empty array. Is that also an invalid policy?

Reply:
1. `{ name:"...", allow_rules:[] }` is invalid due to empty allow rules
2. `{ name:"...", allow_rules:[{name:"allow_all"}] }` is valid
3. `{ name:"...", deny_rules:[...] }` is invalid due to missing allow_rules

In the question, you are referring to the first case right? that should be invalid. But I realized the C++ code does not consider that case, and ends up creating an allow authorization engine with empty policies. So during evaluation, since there is no match, allow engine will always return deny decision. Thank you for bringing this up. Will update the code to return invalid for first case.